lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -53,6 +53,21 @@ def find(request)
     node.trusted_data = Puppet.lookup(:trusted_information) { Puppet::Context::TrustedInformation.local(node) }.to_h
 
     if node.environment
+      # If the request is from a remote node, then the request's environment is
+      # the agent's current environment, the one it last pluginsynced in. If it
+      # doesn't match the server-specified enviroment set by the node terminus,
+      # usually an ENC/classifier, then return an empty catalog specifying the
+      # server specified environment. The agent will detect the mismatch and
+      # retry pluginsync and catalog requests using the server specified
+      # environment. This way both pluginsync and compilation use the same
+      # environment. If we don't do this, then compilation can fail if puppet
+      # code in the server specified environment references facts that don't
+      # exist in the agent's current environment. Or if the fact versions are
+      # different, then compilation can produce undefined results.
+      if request.remote? && request.environment.name != :'*root*' && node.environment != request.environment
+        return Puppet::Resource::Catalog.new(node.name, node.environment)
+      end
```

Comment:
We probably need to make this opt-in based on an HTTP query parameter which can be checked here via `request.options`